src/browser/webgl.rs

```diff
@@ -0,0 +1,332 @@
+      .map_err(JsValueError)?
+      .ok_or("Failed to retrieve context")?
+      .cast::<WebGl2RenderingContext>()?;
+
+    context.viewport(
+      0,
+      0,
+      canvas.width().try_into()?,
+      canvas.height().try_into()?,
+    );
+
+    let program = Self::create_program(
+      &context,
+      vec![
+        ShaderDescription {
+          code: VERTEX.into(),
+          shader_type: WebGl2RenderingContext::VERTEX_SHADER,
+        },
+        ShaderDescription {
+          code: FRAGMENT.into(),
+          shader_type: WebGl2RenderingContext::FRAGMENT_SHADER,
+        },
+      ],
+    )?;
+
+    let length = Self::setup_triangles(&context, &program)?;
+
+    let textures = (0..2)
+      .map(|_| Self::create_texture(&context))
+      .collect::<Result<Vec<WebGlTexture>, _>>()?;
+
+    let frame_buffer = context
+      .create_framebuffer()
+      .ok_or("Failed to create frame buffer")?;
+
+    Ok(Self {
+      context,
+      frame_buffer,
+      length,
+      program,
+      source: Cell::new(0),
+      textures,
+    })
+  }
+
+  pub(crate) fn render_to_canvas(&self, state: &Computer) -> Result {
+    self
+      .context
+      .bind_framebuffer(WebGl2RenderingContext::FRAMEBUFFER, None);
+
+    self.context.bind_texture(
+      WebGl2RenderingContext::TEXTURE_2D,
+      Some(&self.textures[self.source.get()]),
+    );
+
+    self.context.uniform1i(
+      self
+        .context
+        .get_uniform_location(&self.program, &state.mask().to_string())
+        .as_ref(),
+      0,
+    );
+
+    self.context.uniform1i(
+      self
+        .context
+        .get_uniform_location(&self.program, &state.operation().to_string())
+        .as_ref(),
+      0,
+    );
+
+    self.context.draw_arrays(
+      WebGl2RenderingContext::TRIANGLES,
+      0,
+      self.length.try_into()?,
+    );
+
+    Ok(())
+  }
+
+  pub(crate) fn render_to_texture(&self, state: &Computer) -> Result {
+    self.context.bind_framebuffer(
+      WebGl2RenderingContext::FRAMEBUFFER,
+      Some(&self.frame_buffer),
+    );
+
+    self.context.framebuffer_texture_2d(
+      WebGl2RenderingContext::FRAMEBUFFER,
+      WebGl2RenderingContext::COLOR_ATTACHMENT0,
+      WebGl2RenderingContext::TEXTURE_2D,
+      Some(&self.textures[self.source.get() ^ 1]),
+      0,
+    );
+
+    self.context.bind_texture(
+      WebGl2RenderingContext::TEXTURE_2D,
+      Some(&self.textures[self.source.get()]),
+    );
+
+    self.context.uniform1i(
+      self
+        .context
+        .get_uniform_location(&self.program, &state.mask().to_string())
+        .as_ref(),
+      1,
+    );
+
+    self.context.uniform1i(
+      self
+        .context
+        .get_uniform_location(&self.program, &state.operation().to_string())
+        .as_ref(),
+      1,
+    );
+
+    self.context.draw_arrays(
+      WebGl2RenderingContext::TRIANGLES,
+      0,
+      self.length.try_into()?,
+    );
+
+    self.source.set(self.source.get() ^ 1);
+
+    Ok(())
+  }
+
+  fn create_program(
```

Comment:
This is only ever called with a vertex shader and a fragment shader, so I think `ShaderDescription` is unnecessary, and it can just take the vertex shader source and the fragment shader source as two different arguments. If we eventually have more complicated programs, we can add something like this back.